File: carnerf-mobile/lib/utils/week.ts
// KST(UTC+9) 기준 ISO 주차 키 생성. 매주 월요일 0시(KST)에 키가 바뀐다.
const KST_OFFSET_MS = 9 * 60 * 60 * 1000;

export function nowKst(): Date {
  return new Date(Date.now() + KST_OFFSET_MS);
}

// 주 시작(월요일 0시 KST)의 timestamp 반환
export function weekStartKst(d: Date = nowKst()): Date {
  const dow = d.getUTCDay();
  const diff = (dow + 6) % 7;
  const ws = new Date(d);
  ws.setUTCDate(ws.getUTCDate() - diff);
  ws.setUTCHours(0, 0, 0, 0);
  return ws;
}

// ISO 주차 키: "2026-W20"
export function weekKey(d: Date = nowKst()): string {
  const ws = weekStartKst(d);
  const year = ws.getUTCFullYear();
  const first = new Date(Date.UTC(year, 0, 1));
  const days = Math.floor((ws.getTime() - first.getTime()) / 86400000);
  const week = Math.ceil((days + first.getUTCDay() + 1) / 7);
  return `${year}-W${String(week).padStart(2, '0')}`;
}

// 다음 월요일까지 남은 일/시/분/초
export function timeToNextMondayKst(d: Date = nowKst()): {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
  totalMs: number;
} {
  const ws = weekStartKst(d);
  const next = new Date(ws);
  next.setUTCDate(next.getUTCDate() + 7);
  const totalMs = next.getTime() - d.getTime();
  return {
    days: Math.floor(totalMs / 86400000),
    hours: Math.floor((totalMs % 86400000) / 3600000),
    minutes: Math.floor((totalMs % 3600000) / 60000),
    seconds: Math.floor((totalMs % 60000) / 1000),
    totalMs,
  };
}

// 결정론적 PRNG (mulberry32)
export function mulberry32(seed: number) {
  let a = seed >>> 0;
  return function () {
    a |= 0;
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

// 문자열 → 정수 시드 (FNV-1a)
export function hashSeed(s: string): number {
  let h = 2166136261 >>> 0;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0;
}

File: carnerf-flutter/lib/core/utils/week.dart
const int _kstOffsetMs = 9 * 60 * 60 * 1000;

DateTime nowKst() =>
    DateTime.fromMillisecondsSinceEpoch(DateTime.now().millisecondsSinceEpoch + _kstOffsetMs, isUtc: true);

DateTime weekStartKst([DateTime? d]) {
  final base = d ?? nowKst();
  final dow = base.weekday;
  final diff = (dow + 6) % 7;
  final ws = DateTime.utc(base.year, base.month, base.day).subtract(Duration(days: diff));
  return ws;
}

String weekKey([DateTime? d]) {
  final ws = weekStartKst(d);
  final year = ws.year;
  final first = DateTime.utc(year, 1, 1);
  final days = ws.difference(first).inDays;
  final week = ((days + first.weekday) / 7).ceil();
  return '$year-W${week.toString().padLeft(2, '0')}';
}

class CountdownParts {
  const CountdownParts({
    required this.days,
    required this.hours,
    required this.minutes,
    required this.seconds,
    required this.totalMs,
  });

  final int days;
  final int hours;
  final int minutes;
  final int seconds;
  final int totalMs;
}

CountdownParts timeToNextMondayKst([DateTime? d]) {
  final base = d ?? nowKst();
  final ws = weekStartKst(base);
  final next = ws.add(const Duration(days: 7));
  final totalMs = next.difference(base).inMilliseconds;
  return CountdownParts(
    days: totalMs ~/ 86400000,
    hours: (totalMs % 86400000) ~/ 3600000,
    minutes: (totalMs % 3600000) ~/ 60000,
    seconds: (totalMs % 60000) ~/ 1000,
    totalMs: totalMs,
  );
}

double Function() mulberry32(int seed) {
  var a = seed & 0xFFFFFFFF;
  return () {
    a = (a + 0x6D2B79F5) & 0xFFFFFFFF;
    var t = a;
    t = (t ^ (t >>> 15)) & 0xFFFFFFFF;
    t = _imul(t, 1 | a);
    t = (t + _imul(t ^ (t >>> 7), 61 | t)) & 0xFFFFFFFF;
    t = t ^ t;
    t = a;
    t = (t ^ (t >>> 15)) & 0xFFFFFFFF;
    t = _imul(t, 1 | a);
    var u = (t ^ (t >>> 7)) & 0xFFFFFFFF;
    u = _imul(u, 61 | t);
    t = (t + u) & 0xFFFFFFFF;
    t = (t ^ (t >>> 14)) & 0xFFFFFFFF;
    return t / 4294967296;
  };
}

int _imul(int a, int b) {
  final aHi = (a >>> 16) & 0xFFFF;
  final aLo = a & 0xFFFF;
  final bHi = (b >>> 16) & 0xFFFF;
  final bLo = b & 0xFFFF;
  return ((aLo * bLo) + (((aHi * bLo + aLo * bHi) & 0xFFFF) << 16)) & 0xFFFFFFFF;
}

int hashSeed(String s) {
  var h = 2166136261 & 0xFFFFFFFF;
  for (final code in s.codeUnits) {
    h = (h ^ code) & 0xFFFFFFFF;
    h = _imul(h, 16777619);
  }
  return h & 0xFFFFFFFF;
}
